client/index.js

```diff
@@ -626,122 +558,7 @@ PersonaClient.prototype.isPresignedUrlValid = function (presignedUrl, secret) {
-                            } else if (data.access_token) {
-                                // cache token
-                                var cacheFor = data.expires_in - 60, // cache for token validity minus 60s
-                                    now = (new Date().getTime() / 1000);
-                                data.expires_at = now + data.expires_in;
-                                if (cacheFor > 0) {
-                                    _this.tokenCache.set(cacheKey, JSON.stringify(data), cacheFor);
-                                    callback(null, data);
-                                } else {
-                                    callback(null, data);
-                                }
-                            } else {
-                                callback("Could not get access token", null);
-                            }
-                        });
-                    } else {
-                        var err = "Generate token failed with status code " + resp.statusCode;
-                        _this.error(err);
-                        callback(err, null);
-                    }
-                });
-                req.on("error", function (e) {
-                    var err = "OAuth::generateToken problem: " + e.message;
-                    _this.error(err);
-                    callback(err, null);
-                });
-                req.write(post_data);
-                req.end();
-            } else {
-                var data;
-                if (_.isObject(reply)) {
-                    data = reply;
-                    reply = JSON.stringify(data);
-                } else {
-                    data = JSON.parse(reply);
-                }
-                _this.debug("Found cached token for key " + cacheKey + ": " + reply);
-
-                if (data.access_token) {
-                    // recalc expires_in
-                    var now = new Date().getTime() / 1000;
-                    var expiresIn = data.expires_at - now;
-                    _this.debug("New expires_in is " + expiresIn);
-                    if (expiresIn > 0) {
-                        data.expires_in = expiresIn;
-                        callback(null, data);
-                        return;
-                    }
-                }
-                // either expiresIn<=0, or malformed data, remove from redis and retry
-                _this._removeTokenFromCache(id, secret, function (err) {
-                    if (err) {
-                        callback(err, null);
-                    } else {
-                        // iterate
-                        _this.obtainToken({
-                            id: id,
-                            secret: secret,
-                            xRequestId: xRequestId
-                        }, callback);
-                    }
-                });
-            }
-        }
-    }.bind(this));
+    publicKeyStore.obtainToken(this, opts, callback);
```

Comment:
Not sure what I think of handing in `this` as it tightly bounds the client to the methods being called, it feels a bit odd.

Reply:
Its actually a fairly common way of doing things in nodes. It makes functions like `obtainToken` really easy to unit test by stubbing out the client.

Reply:
I guess what I dislike is not knowing what the client should look like.